Python/optimizer_analysis.c

```diff
@@ -533,6 +533,128 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
     Py_UNREACHABLE();
 }
 
+/* _PUSH_FRAME/_POP_FRAME's operand can be 0, a PyFunctionObject *, or a
+ * PyCodeObject *. Retrieve the code object if possible.
+ */
+static PyCodeObject *
+get_co(_PyUOpInstruction *op) {
+    assert(op->opcode == _PUSH_FRAME || op->opcode == _POP_FRAME);
+    PyCodeObject *co = NULL;
+    uint64_t operand = op->operand;
+    if (operand == 0) {
+        return NULL;
+    }
+    if (operand & 1) {
+        co = (PyCodeObject *)(operand & ~1);
+    }
+    else {
+        PyFunctionObject *func = (PyFunctionObject *)operand;
+        assert(PyFunction_Check(func));
+        co = (PyCodeObject *)func->func_code;
+    }
+    assert(PyCode_Check(co));
+    return co;
+}
+
+// _PUSH_FRAME is 3 ops in front of _CHECK_STACK_SPACE
+#define _CHECK_STACK_TO_PUSH_FRAME_OFFSET 3
+
+/* Compute the maximum stack space needed for this trace and consolidate
+ * all possible _CHECK_STACK_SPACE ops to one _CHECK_STACK_SPACE_OPERAND.
+ * This calculation is essentially a traversal over the tree of function
+ * calls in this trace.
+ */
+static void
+combine_stack_space_checks(
+    _PyUOpInstruction *buffer,
+    int buffer_size
+)
+{
+#ifdef Py_DEBUG
+    // assume there's a _PUSH_FRAME after every _CHECK_STACK_SPACE
+    bool expecting_push = false;
+    // if we ever can't get a framesize from _PUSH_FRAME, must exit soon
+    bool must_exit = false;
+#endif
+    int depth = 0;
+    uint32_t space_at_depth[MAX_ABSTRACT_FRAME_DEPTH];
```

Comment:
You only need to track the current depth, as this is a single pass algorithm.
In fact, this may be incorrect, if more than one function is called with differing stack requirements.

I think, only the following variables are needed:
* The first check (address or offset)
* The current stack size
* The maximum requested stack size

Reply:
Plus the last seen stack check, to avoid hard-coding the offset between `CHECK_STACK_SPACE` and `_PUSH_FRAME`.

Reply:
Thanks for the suggestion - I was able to remove my makeshift stack. I had to use one additional variable to keep track of the current frame's size, since there's no way I could see to access this from `_POP_FRAME` (we only have the code object being returned to, not the one we're returning from)